shadowsocks/udp.go

```diff
@@ -72,36 +73,43 @@ func unpack(clientIP net.IP, dst, src []byte, cipherList CipherList) ([]byte, st
 }
 
 type udpService struct {
-	clientConn     net.PacketConn
-	natTimeout     time.Duration
-	ciphers        CipherList
-	m              metrics.ShadowsocksMetrics
-	isRunning      bool
-	checkAllowedIP func(net.IP) *onet.ConnectionError
+	clientConn net.PacketConn
+	running    sync.WaitGroup
 }
 
 // NewUDPService creates a UDPService
-func NewUDPService(clientConn net.PacketConn, natTimeout time.Duration, cipherList CipherList, m metrics.ShadowsocksMetrics) UDPService {
-	return &udpService{clientConn: clientConn, natTimeout: natTimeout, ciphers: cipherList, m: m, checkAllowedIP: onet.RequirePublicIP}
+func NewUDPService(clientConn net.PacketConn, natTimeout time.Duration, cipherList CipherList, m metrics.ShadowsocksMetrics, ipPolicy onet.IPPolicy) UDPService {
```

Comment:
Same as TCP: let's separate the construction and the serve call.

Reply:
Done